tests/test_autosave.py

```diff
@@ -0,0 +1,492 @@
+        saved = yaml.full_load(f)
+    assert saved["aOut"] == 20.0
+    assert saved["aIn"] == 20.0
+    assert saved["boolOut"] == 1
+    assert saved["boolIn"] == 1
+    assert saved["longIn"] == 20
+    assert saved["longOut"] == 20
+    assert saved["int64In"] == 100
+    assert saved["int64Out"] == 100
+    assert saved["mbbIn"] == 15
+    assert saved["mbbOut"] == 15
+    assert saved["stringIn"] == "test string in"
+    assert saved["stringOut"] == "test string out"
+    assert saved["longStringIn"] == "test long string in"
+    assert saved["longStringOut"] == "test long string out"
+    assert saved["Action"] == 1
+    assert (saved["WaveformIn"] == numpy.array([1, 2, 3, 4])).all()
+    assert (saved["WaveformOut"] == numpy.array([1, 2, 3, 4])).all()
+    autosave.Autosave._stop()
+    # force autosave thread to stop to ensure pytest exits
+    conn.send("D")
+
+
+def test_actual_ioc_save(tmp_path):
+    ctx = get_multiprocessing_context()
+    parent_conn, child_conn = ctx.Pipe()
+    ioc_process = ctx.Process(
+        target=check_all_record_types_save_properly,
+        args=(DEVICE_NAME, tmp_path, child_conn),
+    )
+    ioc_process.start()
+    # If we never receive D it probably means an assert failed
+    select_and_recv(parent_conn, "D")
+
+
+def check_autosave_field_names_contain_device_prefix(
+    device_name, tmp_path, conn
+):
+    autosave.configure(tmp_path, device_name, save_period=1)
+    builder.aOut("BEFORE", autosave=["VAL", "EGU"])
+    builder.SetDeviceName(device_name)
+    builder.aOut("AFTER", autosave=["VAL", "EGU"])
+    builder.LoadDatabase()
+    softioc.iocInit()
+    time.sleep(2)
+    with open(tmp_path / f"{device_name}.softsav", "r") as f:
+        saved = yaml.full_load(f)
+    assert "BEFORE" in saved.keys()
+    assert f"{device_name}:AFTER" in saved.keys()
+    autosave.Autosave._stop()
+    conn.send("D")
+
+
+def test_autosave_field_names_contain_device_prefix(tmp_path):
+    ctx = get_multiprocessing_context()
+    parent_conn, child_conn = ctx.Pipe()
+    ioc_process = ctx.Process(
+        target=check_autosave_field_names_contain_device_prefix,
+        args=(DEVICE_NAME, tmp_path, child_conn),
+    )
+    ioc_process.start()
+    # If we never receive D it probably means an assert failed
+    select_and_recv(parent_conn, "D")
+
+def test_context_manager_thread_safety(tmp_path):
+    autosave.configure(tmp_path, DEVICE_NAME)
+
+    def create_pv_in_thread(name, wait):
+        time.sleep(wait)
+        builder.aOut(name, autosave=False)
+
+    pv_thread_before_cm = threading.Thread(
+        target=create_pv_in_thread, args=["PV-FROM-THREAD-BEFORE", 1])
+    pv_thread_in_cm = threading.Thread(
+        target=create_pv_in_thread, args=["PV-FROM-THREAD-DURING", 0])
+    pv_thread_before_cm.start()
+    with autosave.Autosave(["VAL", "EGU"]):
+        builder.aOut("PV-FROM-CM")
+        pv_thread_in_cm.start()
+        pv_thread_in_cm.join()
+    pv_thread_before_cm.join()
```

Comment:
On initial impression this test looks pretty flaky - relying on a `time.sleep` is not stable, especially across different platforms which have different threading behaviours. I've observed the MacOS GH runners taking significantly longer to spin up threads than other platforms, for example.

A better way would be to create a couple of `threading.Event` objects in the main level, and have the child threads wait on them until the main thread sets them. That will ensure much more consistency.